### PR TITLE
fixed issue #6 and #8

### LINKED ISSUE #6: Switches from non-numbered workspaces to first numbered workspace after rename is called.
If a named workspace is visible on one output, and the rename function is called via a change on another output, refocusing back to the named workspace causes that output to switch to the first numbered workspace on that output. This does not occur with the daemon disabled, but does occur with the version prior to [Merge pull request #5 from kensand/master](https://github.com/cboddy/i3-workspace-names-daemon/tree/181c73ea5199eac9ff6d413a13139318013153e5), so I am fairly certain it is script related but not related to the handling of named workspaces. Perhaps something to do with a side effect of renaming all windows in i3? I will look into it further soon, as this script is great, but this bug is severe for me.

For reference - Arch x86_64 Linux 4.18.16-arch1-1-vfio - i3-gaps - polybar - ttf-font-awesome-4 (AUR)

### PR DESCRIPTION
**Pull request by kensand**
This fixes  #6 by saving the active and focused workspaces prior to renaming them, then re-activates and re-focuses them using the new names assigned to them. Also combines all i3 commands into a single string so that order of activation is preserved, otherwise the wrong workspace may be focused at the end of the rename function. Thanks again for the awesome script @cboddy - it's fantastic for me now that #6 is fixed. I'll be happy to continue assist in maintaining in the future.

### CODE REVIEW

**Comment by kensand** [comment]
fixed issue number 8 in the second commit by using the leaf name in the event that the leaf does not have a class.

File: i3_workspace_names_daemon.py [hunk 1]
```diff
@@ -39,21 +38,55 @@ def build_rename(i3, app_icons, delim):
         The rename callback.
     """
     def get_icon_or_name(leaf):
-        window_class = leaf.window_class.lower()
+        window_class = ""
```
Comment (cboddy) [inline i3_workspace_names_daemon.py]: this statement can be removed.

File: i3_workspace_names_daemon.py [hunk 2]
```diff
@@ -39,21 +38,55 @@ def build_rename(i3, app_icons, delim):
...
             return window_class.lower()
 
     def rename(i3, e):
-        for workspace in i3.get_tree().workspaces():
+        workspaces = i3.get_tree().workspaces()
+        #need to use get_workspaces since the i3 con object doesn't have the visible property for some reason
+        workdicts = i3.get_workspaces()
+        visible =  [workdict['name'] for workdict in workdicts if workdict['visible']]
+        visworkspaces = []
+        focus = [workdict['name'] for workdict in workdicts if workdict['focused']]
```
Comment (cboddy) [inline i3_workspace_names_daemon.py]: more concise way to get the first element:
```
import itertools

focus = next(itertools.chain([workdict['name'] for workdict in workdicts if workdict['focused']], [None])
```

File: i3_workspace_names_daemon.py [hunk 3]
```diff
@@ -39,21 +38,55 @@ def build_rename(i3, app_icons, delim):
...
+            focus = focus[0]
+        else:
+            focus = None
+
+        commands = []
+        for workspace in workspaces:
             names = [get_icon_or_name(leaf)
                      for leaf in workspace.leaves()]
             names = delim.join(names)
+            newname = ""
```
Comment (cboddy) [inline i3_workspace_names_daemon.py]: this statement can be removed.

File: i3_workspace_names_daemon.py [hunk 4]
```diff
@@ -39,21 +38,55 @@ def build_rename(i3, app_icons, delim):
...
+                visworkspaces.append(newname)
+            if workspace.name == focus:
+                focusname = newname
+
+            commands.append('rename workspace "{}" to "{}"'.format(workspace.name, newname))
+
+        for workspace in visworkspaces + [focusname]:
+            commands.append('workspace {}'.format(workspace))
+
+        #we have to join all the activate workspaces commands into one or the order
```
Comment (cboddy) [inline i3_workspace_names_daemon.py:80]: ah, makes sense.

File: i3_workspace_names_daemon.py [hunk 5]
```diff
@@ -39,21 +38,55 @@ def build_rename(i3, app_icons, delim):
...
+
+            if workspace.name in visible:
+                visworkspaces.append(newname)
+            if workspace.name == focus:
+                focusname = newname
+
+            commands.append('rename workspace "{}" to "{}"'.format(workspace.name, newname))
+
+        for workspace in visworkspaces + [focusname]:
+            commands.append('workspace {}'.format(workspace))
```
Comment (cboddy) [inline i3_workspace_names_daemon.py:78]: why is this necessary?

**Review by cboddy — CHANGES_REQUESTED** [review]
LGTM, a few minor suggestions (and one question to reduce my general ignorance).

I'd like to run with this version for a day or two before merging (but after the next update). If any of the suggestions are unclear or you'd prefer I do them let me know.

**Comment by cboddy** [comment]
> This fixes #6 by saving the active and focused workspaces prior to renaming them, then re-activates and re-focuses them using the new names assigned to them. Also combines all i3 commands into a single string so that order of activation is preserved, otherwise the wrong workspace may be focused at the end of the rename function. Thanks again for the awesome script @cboddy - it's fantastic for me now that #6 is fixed. I'll be happy to continue assist in maintaining in the future.

@kensand thanks for taking the time to do this, your help is most welcome!

File: i3_workspace_names_daemon.py (hunk 1 above)
Comment (kensand) [reply i3_workspace_names_daemon.py]: Yes, it can be.

File: i3_workspace_names_daemon.py (hunk 2 above)
Comment (kensand) [reply i3_workspace_names_daemon.py]: True, albeit a little hard to read in my opinion. Will change.

File: i3_workspace_names_daemon.py (hunk 3 above)
Comment (kensand) [reply i3_workspace_names_daemon.py]: Will do.

File: i3_workspace_names_daemon.py (hunk 4 above)
Comment (kensand) [reply i3_workspace_names_daemon.py:80]: Yeah this is the solution to #6. I assume there is some underlying async in i3ipc that can allow the order of command to be changed. The simple solution is to just issue one long command at the end of rename. I should probably check and make sure there is no command length limit in i3ipc or i3 itself.

File: i3_workspace_names_daemon.py (hunk 2 above)
Comment (kensand) [reply i3_workspace_names_daemon.py]: The line you suggested wont work because it makes a 2D array. The following does work however:
`focus = ([workdict['name'] for workdict in workdicts if workdict['focused']] or [None])[0]`

**Comment by kensand** [comment]
sorry, missed one change in that last commit, will fix in just a sec

File: i3_workspace_names_daemon.py (hunk 2 above)
Comment (kensand) [reply i3_workspace_names_daemon.py]: @cboddy I see you updated your comment. Would you prefer you're way or mine? I personally would opt for mine since I try not to import when it can be avoided.

Comment (cboddy) [reply i3_workspace_names_daemon.py]: @kensand NP, yours looks good.

**Comment by kensand** [comment]
I believe I changed everything you requested. Sorry for some silly mistakes, and thanks for going through my code. Let me know if there's anything else you'd like changed!

File: i3_workspace_names_daemon.py (hunk 2 above)
Comment (kensand) [reply i3_workspace_names_daemon.py]: Uh oh... my way is causing focus to be None incorrectly for some reason, I'll try you're way

**Comment by kensand** [comment]
It looks like i3 messages are limited to a max length of 2^32 bytes or ~4 gigs according to section 2 of https://i3wm.org/docs/ipc.html so I don't think command message length will be a concern.

**Comment by cboddy** [comment]
Thanks for checking that @kensand .
I've been testing (running i3 with) this version for the past few days without issue so I think it's good to go, thanks!
